core/src/main/java/hudson/util/RobustCollectionConverter.java

```diff
@@ -85,9 +108,19 @@ protected void populateCollection(HierarchicalStreamReader reader, Unmarshalling
             reader.moveDown();
             try {
                 Object item = readBareItem(reader, context, collection);
-                long nanoNow = System.nanoTime();
-                collection.add(item);
-                XStream2SecurityUtils.checkForCollectionDoSAttack(context, nanoNow);
+                try {
+                    if (elementType != null) {
+                        // When possible, disallow invalid elements.
+                        elementType.cast(item);
+                    }
+                    long nanoNow = System.nanoTime();
+                    collection.add(item);
+                    XStream2SecurityUtils.checkForCollectionDoSAttack(context, nanoNow);
+                } catch (ClassCastException e) {
```

Comment:
Can this be thrown from something after the `Class.cast`? If not, it would be better to use `isInstance` without a `try`-`catch` block.

Reply:
Adjusted in [df2ac81](https://github.com/jenkinsci/jenkins/pull/9727/commits/df2ac81257510f5f0ac715d2bf5036b20b2ffaf7).